core/src/main/scala/caliban/schema/ArgBuilder.scala

```diff
@@ -215,6 +215,27 @@ trait ArgBuilderInstances extends ArgBuilderDerivation {
     case other                       => ev.build(other).map(List(_))
   }
 
+  implicit def map[K, V](implicit keyEv: ArgBuilder[K], valueEv: ArgBuilder[V]): ArgBuilder[Map[K, V]] = {
+    case InputValue.ListValue(kvs) =>
+      kvs
+        .foldLeft[Either[ExecutionError, Map[K, V]]](Right(Map.empty)) {
+          case (res @ Left(_), _)                       => res
+          case (Right(res), InputValue.ObjectValue(kv)) =>
+            for {
+              key   <- if (kv.contains("key")) { keyEv.build(kv("key")) }
+                       else { Left(InvalidInputArgument("key", kv)) }
+              value <-
+                if (kv.contains("value")) {
+                  valueEv.build(kv("value"))
+                } else {
+                  Left(InvalidInputArgument("value", kv))
+                }
+            } yield res + (key -> value)
```

Comment:
If you continue to use an immutable map, prefer `updated` to avoid the tuple allocation.
```suggestion
            } yield res.updated(key, value)
```